### PR TITLE
Improve misleading ORMInvalidArgumentException message

### PR DESCRIPTION
**Pull request by jonny-no1**
See https://github.com/doctrine/doctrine2/issues/6029#issuecomment-340304346.

### CODE REVIEW

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+/** @Entity */
+class DDC6029Group implements DDC6029GroupInterface
+{
+    /** @Id @Column(type="integer") @GeneratedValue */
+    public $id;
+
+    /**
+     * @ManyToMany(targetEntity="DDC6029User", mappedBy="groups")
+     */
+    private $users;
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: Is this needed? Let's keep only the things required by the tests

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        ]);
+    }
+
+    /**
+     * Verifies that when wrong entity is persisted via relationship field, the error message does not correctly state
+     * the expected class name.
+     */
+    public function testIssue()
+    {
+        $user = new DDC6029User();
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: Shouldn't we add here the expected exception and the expected exception message? Like:

```php
$this->expectException(ORMInvalidArgumentException::class);
$this->expectExceptionMessage('Expected value of type "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group" for association field "Doctrine\Tests\ORM\Functional\Ticket\DDC6029User#$groups", got "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group2" instead.');
```

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            $this->_em->getClassMetadata(DDC6029Group::class),
+            $this->_em->getClassMetadata(DDC6029Group2::class),
+        ]);
+    }
+
+    /**
+     * Verifies that when wrong entity is persisted via relationship field, the error message does not correctly state
+     * the expected class name.
+     */
+    public function testIssue()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: ` : void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        $this->_schemaTool->createSchema([
+            $this->_em->getClassMetadata(DDC6029User::class),
+            $this->_em->getClassMetadata(DDC6029Group::class),
+            $this->_em->getClassMetadata(DDC6029Group2::class),
+        ]);
+    }
+
+    /**
+     * Verifies that when wrong entity is persisted via relationship field, the error message does not correctly state
+     * the expected class name.
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: Please add `@group` with the ticket number so that we can trace things back

**Review by lcobucci — CHANGES_REQUESTED** [review]

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public function testIssue(): void
+    {
+        $user = new DDC6029User();
+        $user->addGroup(new DDC6029Group2());
+
+        self::expectException(ORMInvalidArgumentException::class);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: This method is not static, so we should use `$this` instead

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        self::expectExceptionMessage('Expected value of type "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group" for association field "Doctrine\Tests\ORM\Functional\Ticket\DDC6029User#$groups", got "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group2" instead.');
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php]: This method is not static, so we should use `$this` instead

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 7]
```diff
@@ -195,11 +195,7 @@ public static function invalidIdentifierBindingEntity()
      */
     public static function invalidAssociation(ClassMetadata $targetClass, $assoc, $actualValue)
     {
-        $expectedType = 'Doctrine\Common\Collections\Collection|array';
-
-        if (($assoc['type'] & ClassMetadata::TO_ONE) > 0) {
```
Comment (lcobucci) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: Is there any explanation on why we had this code there (commit msg)? Maybe @Ocramius can give us some context

**Review by lcobucci — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 7 above)
Comment (foaly-nr1) [reply lib/Doctrine/ORM/ORMInvalidArgumentException.php]: No details in the commit message of https://github.com/doctrine/doctrine2/commit/71a6a88de82d23232cdf5377a44804a5f55d29b8 and not sure where I could find details on DDC-3490.

**Review by lcobucci — APPROVED** [review]
I've rebased and added a test for one-to-many association as well.

**Comment by lcobucci** [comment]
@foaly-nr1 🚢 thanks for your contribution!

**Pull request merged**